### ISSUE TITLE
Column key is regenerated automatically even if key is already defined

### ISSUE THREAD

**Issue by neospyk** [schrodinger/fixed-data-table-2#84]
Hello,
Now I am implementing hidden columns using your example. And I've noticed that after removing one column, next columns keys were regenerated.

## Expected Behavior
Column keys should not be regenerated on each render.

## Current Behavior
Column keys are generated on each render. 

## Possible Solution
I have tried allowCellsRecycling=true but after removing column keys were regenerated because next columns move one position left.
I'd propose to update component without key regeneration or/and provide separate key parameter in Column component.
https://github.com/schrodinger/fixed-data-table-2/blob/master/src/FixedDataTableCellGroup.js
```javascript
render() {
...
for (var i = 0, j = columns.length; i < j; i++) {
      var columnProps = columns[i].props;
      var recycable = columnProps.allowCellsRecycling && !isColumnReordering;
      if (!recycable || (
            currentPosition - props.left <= props.width &&
            currentPosition - props.left + columnProps.width >= 0)) {
        var key = 'cell_' + i;
        cells[i] = this._renderCell(
          props.rowIndex,
          props.rowHeight,
          columnProps,
          currentPosition,
          key,
          contentWidth,
          isColumnReordering
        );
      }
      currentPosition += columnProps.width;
    }
...
}
```

## Context
I created table with header filters and hidden columns functionality. After filters were applied I started to hide column - all columns after that were recreated as a result filter's state was cleared.

## Your Environment
* Version used: ^0.7.6

**Comment by neospyk** [comment]
I have found step to reproduce:
When we have horizontal scroll and column moves outside of the veiw and then comes back becomes recreated with lose his state. Even with prop allowCellsRecycling=true.

**Comment by wcjordan** [comment]
Hmm this does seem like it could lead to issues.  We could use the columnKey prop as the key instead of a synthetic one, but that might cause other issues if folks aren't assuming columnKey needs to be unique...

Could you create a JsFiddle or Plunker which demonstrates the issue?  Your description isn't explicit, but it seems you may have uncontrolled inputs in your cells.  You should be aware that even with this fix, those inputs may not work as expected and may lose their state while you're scrolling.

**Comment by neospyk** [comment]
Yes, we use uncontrolled input, and we can't use other in header cell for our goals. But when our input in header cell will our of the view and return we lose our data because filter component recreated.

Yes using columnKey it simply way for both, but cause some problem with uniquie key. But of you could provide new props for example "use columnKeyAsKey=true" and discribe it in documentation users will know that columnKey should be uniquie.

Also updating the component instead recreated may inrease performance.

I can show example close to our problem in few hour in JSFiddle

**Comment by neospyk** [comment]
Unfortunately in JSFiddle I can reproduce issue only with scroll. My filter component not so simple    . There is other problem which absent in my code but i think it is kind of the my problem. If you put something to first input and scroll to the end and return input will empty. With hidding the same behavior with all cell after hidded, each other will recreated. And playing with allowCellsRecycling  true/false have not result for hiding, but yes it fix scroll.
Pretty way use it mount/unmount hidded component and and only update each after hidded but it possible only if we will have the same key always for this cell
[JSFiddle](https://jsfiddle.net/y1L3tog3/)

**Comment by neospyk** [comment]
About JSFiddle: to the header I added input field, our to the first input some text, and scroll stable horizontal to the last columnKey, and return, after return to the first column, input will empty. It is expected behavior for current implementation but we can avoid.

And sorry, but what does mean "submit PR"?

**Comment by wcjordan** [comment]
Submit a PR means to create a Pull Request on github with the fix.

Here's the behavior I see in the fiddle.  The input isn't lost.  Maybe my screen resolution is too large?
![jsfiddle](https://cloud.githubusercontent.com/assets/1034455/20642492/466de87a-b3de-11e6-97f9-09123cc62c87.gif)

**Comment by neospyk** [comment]
Ohh. Sorry it was old JSFiddle link. [It's correct JSFiddle](https://jsfiddle.net/jt7b2z1o/). Changed only width. 200 px only for fast reproduce. We have a lot of column with width 130px in full screen table with scroll and have this problem. 
Yes, I will create PR at monday or tuesday

**Comment by wcjordan** [comment]
Cool, the fiddle repros for me now.  Thanks for the help, I look forward to the PR.  Let mw know if you have any questions about the code base.

**Comment by KamranAsif** [comment]
Joining this discussion late, but you would need allowCellsRecycling to be _false_ in order to preserve the input fields scrolling left and right

FDT doesn't actually render columns; it renders rows and each cell is given a key of cell_${index}.

I think a simple patch would be to change [this line](https://github.com/schrodinger/fixed-data-table-2/blob/master/src/FixedDataTableCellGroup.js#L85
) to 
```        var key = columnProps.columnKey || 'cell_' + i;```

**Comment by wcjordan** [comment]
@KamranAsif do you think there's any need to worry about users using non unique columnKeys?  Or do you think it's safe to just swap to columnKey in a patch version bump?

**Comment by KamranAsif** [comment]
Hmm columnKey is suppose to be unique according to the API docs. We could add a check to make sure values are unique in DEV mode

**Comment by wcjordan** [comment]
Nice, then let's forgo the prop and make the change Kam suggests.  If you could try it out and confirm it solves your issue, and if so, submit a PR, we'll merge in the solution.  Thanks!

**Comment by neospyk** [comment]
Thanks. Yes will check and submit a little bit later.

**Comment by neospyk** [comment]
This fix work fine for me. I will submit PR

**Comment by neospyk** [comment]
Hi. As I see fix was merged. Can I hope to receive fix in next minor version(0.7.8 as example) and when? Thanks, after that a will close issue.

**Comment by neospyk** [comment]
Hello guys,
I have to reopen the issue because I closed it accidentally.Could you please create minor release as I need to fix this issue on production.
Thank you

**Comment by KamranAsif** [comment]
@neospyk Hey, just released 0.7.8 with your change.
Closing this unless theres anything else

**Comment by neospyk** [comment]
Thanks a lot
